helper/helper_test.go

```diff
@@ -1,12 +1,204 @@
 package helper
 
 import (
+	"bufio"
+	"fmt"
+	"os"
+
 	"github.com/greenplum-db/gpbackup/utils"
+	"golang.org/x/sys/unix"
 
 	. "github.com/onsi/ginkgo/v2"
 	. "github.com/onsi/gomega"
+
+	"github.com/greenplum-db/gpbackup/toc"
+	"github.com/pkg/errors"
+)
+
+var (
+	testDir     = "/tmp/helper_test/20180101/20180101010101"
+	testTocFile = fmt.Sprintf("%s/test_toc.yaml", testDir)
 )
 
+type restoreReaderTestImpl struct {
+	waitCount int
+}
+
+func (r *restoreReaderTestImpl) waitForPlugin() error {
+	r.waitCount++
+	return nil
+}
+
+func (r *restoreReaderTestImpl) positionReader(pos uint64, oid int) error {
+	return nil
+}
+
+func (r *restoreReaderTestImpl) copyData(num int64) (int64, error) {
+	return 1, nil
+}
+
+func (r *restoreReaderTestImpl) copyAllData() (int64, error) {
+	return 1, nil
+}
+
+func (r *restoreReaderTestImpl) closeFileHandle() {
+}
+
+func (r *restoreReaderTestImpl) getReaderType() ReaderType {
+	return "nil"
+}
+
+type helperTestStep struct {
+	restorePipeWriterArgExpect string
+	restorePipeWriterResult    bool
+	skipFileArgTableOid        int
+	skipFileResult             bool
+	comment                    string
+}
+
+type restoreMockHelperImpl struct {
+	currentStep      int
+	started          bool
+	expectedOidBatch []oidWithBatch
+	expectedSteps    []helperTestStep
+
+	pipesCreated map[string]bool
+	pipesOpened  map[string]bool
+
+	restoreData *restoreReaderTestImpl
+}
+
+func (h *restoreMockHelperImpl) isPipeOpened(pipe string) bool {
+	Expect(h.pipesOpened).ToNot(BeNil())
+
+	return h.pipesOpened[pipe]
+}
+
+func (h *restoreMockHelperImpl) isPipeCreated(pipe string) bool {
+	Expect(h.pipesCreated).ToNot(BeNil())
+
+	return h.pipesCreated[pipe]
+}
+
+func (h *restoreMockHelperImpl) makeStep() helperTestStep {
+	if !h.started {
+		h.started = true
+	} else {
+		h.currentStep++
+	}
```

Comment:
may be remove `started` and put `h.currentStep++` before `return ret`?

Reply:
It's not the same, after making the first step I expect currentStep to have 0, to allow further getStep() calls to return the very first step yet.

makeStep is used in the beginning of the mocked getRestorePipeWriter() to provide with a new step information. It starts stepping process if it's not started yet and advances  a step pointer. The values of steps are used further in the mocked getRestorePipeWriter().

Reply:
how about such
```diff
diff --git a/helper/helper_test.go b/helper/helper_test.go
index ee674631..f1acc421 100644
--- a/helper/helper_test.go
+++ b/helper/helper_test.go
@@ -58,7 +58,6 @@ type helperTestStep struct {
 
 type restoreMockHelperImpl struct {
 	currentStep      int
-	started          bool
 	expectedOidBatch []oidWithBatch
 	expectedSteps    []helperTestStep
 
@@ -81,12 +80,7 @@ func (h *restoreMockHelperImpl) isPipeCreated(pipe string) bool {
 }
 
 func (h *restoreMockHelperImpl) makeStep() helperTestStep {
-	if !h.started {
-		h.started = true
-	} else {
-		h.currentStep++
-	}
-
+	h.currentStep++
 	ret := h.getCurStep()
 	fmt.Printf("Step: %s", ret.comment)
 	return ret
@@ -112,6 +106,7 @@ func newHelperTest(batches []oidWithBatch, steps []helperTestStep) *restoreMockH
 	ret.pipesCreated = make(map[string]bool)
 	ret.pipesOpened = make(map[string]bool)
 	ret.restoreData = &restoreReaderTestImpl{}
+	ret.currentStep = -1
 
 	// pre-create pipes as starter does
 	for i := 0; i < *copyQueue; i++ {
```

Reply:
Conceptually it has 2 states - not started and started with index and current step, and the logic express it, having index in valid range.  -1 is an initially invalid value, and as for me, to have invalid index in the program might be rather not the best idea.

Reply:
You may use 0 index as not started, and 1+ index as started

Reply:
It was the first version rejected by the review